diff --git a/userfeedback/scripts/eventlog b/userfeedback/scripts/eventlog
index 104dfb1c02..6d3033fa90 100755
--- a/userfeedback/scripts/eventlog
+++ b/userfeedback/scripts/eventlog
@@ -15,4 +15,5 @@ EVENTLOG_FILE="/var/log/eventlog.txt"
 # Need to do this before user can request it in Chrome.
 # Moved here to keep out of critical boot path
 
-elogtool list --utc >"${EVENTLOG_FILE}" 2>&1 || true
+echo "It is too costly to run elogtool in Whalebook" >"${EVENTLOG_FILE}" 2>&1 || true
+#elogtool list --utc >"${EVENTLOG_FILE}" 2>&1 || true
