diff --git a/src/mainboard/google/trogdor/Kconfig b/src/mainboard/google/trogdor/Kconfig
index e409aeee67..2e7091e1ab 100644
--- a/src/mainboard/google/trogdor/Kconfig
+++ b/src/mainboard/google/trogdor/Kconfig
@@ -44,9 +44,11 @@ config BOARD_SPECIFIC_OPTIONS
 	select SPI_FLASH_GIGADEVICE
 	select SPI_FLASH_WINBOND
 	select MAINBOARD_HAS_CHROMEOS
-	select MAINBOARD_HAS_TPM2 if !BOARD_GOOGLE_BUBS
-	select SPI_TPM if !BOARD_GOOGLE_BUBS
-	select TPM_GOOGLE_CR50 if !BOARD_GOOGLE_BUBS
+        #if !defined(WHALEOS_BUILD)
+	#select MAINBOARD_HAS_TPM2 if !BOARD_GOOGLE_BUBS
+	#select SPI_TPM if !BOARD_GOOGLE_BUBS
+	#select TPM_GOOGLE_CR50 if !BOARD_GOOGLE_BUBS
+        #endif
 
 config VBOOT
 	select EC_GOOGLE_CHROMEEC_SWITCHES if !BOARD_GOOGLE_BUBS
