diff --git a/ml/handwriting.cc b/ml/handwriting.cc
index 1310897725..aad9a2d53b 100644
--- a/ml/handwriting.cc
+++ b/ml/handwriting.cc
@@ -32,6 +32,7 @@ constexpr char kHandwritingLibraryRelativePath[] = "libhandwriting.so";
 
 // Language codes supported by Longform.
 constexpr char kLanguageCodeEn[] = "en";
+constexpr char kLanguageCodeKo[] = "ko";
 constexpr char kLanguageCodeGesture[] = "gesture_in_context";
 
 const base::FilePath kDlcRoot("/run/imageloader");
@@ -57,6 +58,11 @@ HandwritingRecognizerModelPaths GetModelPaths(
       paths.set_fst_lm_path(
           model_path.Append("latin_indy.compact.fst").value());
       paths.set_recospec_path(model_path.Append("latin_indy.pb").value());
+    } else if (language == kLanguageCodeKo) {
+      const base::FilePath ko_path = model_path.Append("ko");
+      paths.set_reco_model_path(ko_path.Append("latin_indy.tflite").value());
+      paths.set_fst_lm_path(ko_path.Append("compact.fst.local").value());
+      paths.set_recospec_path(ko_path.Append("qrnn.recospec.local").value());
     } else if (language == kLanguageCodeGesture) {
       paths.set_reco_model_path(
           model_path.Append("gic.reco_model.tflite").value());
diff --git a/ml/seccomp/ml_service-WebPlatformHandwritingModel-seccomp-amd64.policy b/ml/seccomp/ml_service-WebPlatformHandwritingModel-seccomp-amd64.policy
index c711cccc8a..892856cf33 100644
--- a/ml/seccomp/ml_service-WebPlatformHandwritingModel-seccomp-amd64.policy
+++ b/ml/seccomp/ml_service-WebPlatformHandwritingModel-seccomp-amd64.policy
@@ -7,6 +7,7 @@
 # rseq should be first due to frequency and purpose to decrease nptl latency
 rseq: 1
 
+access: 1
 brk: 1
 clock_gettime: 1
 clock_nanosleep: 1
@@ -20,6 +21,7 @@ exit_group: 1
 fcntl: 1
 flock: 1
 fstat: 1
+futex: 1
 fstatfs: 1
 getcpu: 1
 getpid: 1
@@ -27,6 +29,7 @@ getrandom: arg2 in ~GRND_RANDOM
 gettid: 1
 gettimeofday: 1
 lseek: 1
+lstat: 1
 madvise: arg2 == MADV_DONTNEED
 memfd_create: 1
 mmap: arg2 in ~PROT_EXEC || arg2 in ~PROT_WRITE
@@ -36,6 +39,7 @@ munmap: 1
 newfstatat: 1
 openat: 1
 read: 1
+readlink: 1
 recvmsg: 1
 rt_sigaction: 1
 rt_sigprocmask: 1
@@ -49,3 +53,4 @@ statx: 1
 sysinfo: 1
 uname: 1
 write: 1
+writev: 1
